os/drivers/usbhost/uvc/usbhost_uvc_driver.c

```diff
@@ -0,0 +1,2377 @@
+static inline int usbhost_uvc_devinit(FAR struct usbhost_uvc_state_s *priv);
+
+/* Transfer descriptor memory management */
+static int usbhost_alloc_buffers(FAR struct usbhost_uvc_state_s *priv);
+static void usbhost_free_buffers(FAR struct usbhost_uvc_state_s *priv);
+
+/* struct usbhost_uvc_registry_s methods */
+
+static struct usbhost_class_s *usbhost_uvc_create(FAR struct usbhost_hubport_s *hport, FAR const struct usbhost_id_s *id);
+
+/* struct usbhost_class_s methods */
+
+static int usbhost_uvc_connect(FAR struct usbhost_class_s *usbclass, FAR const uint8_t *configdesc, int desclen);
+static int usbhost_uvc_disconnected(FAR struct usbhost_class_s *usbclass);
+
+/* Driver methods -- depend upon the type of TinyAra driver interface exported */
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+/* This structure provides the registry entry ID information that will  be
+ * used to associate the USB class driver to a connected USB device.
+ */
+
+static const const struct usbhost_id_s g_id = {
+	USB_CLASS_MISC,				/* base -- Must be one of the USB_CLASS_* definitions in usb.h */
+	0x02,						/* subclass -- depends on the device */
+	0x01,						/* proto -- depends on the device    */
+	0,							/* vid      */
+	0							/* pid      */
+};
+
+/* This is the USB host storage class's registry entry */
+
+static struct usbhost_registry_s g_uvc = {
+	NULL,						/* flink    */
+	usbhost_uvc_create,			/* create   */
+	1,							/* nids     */
+	&g_id						/* id[]     */
+};
+
+/* This is the video lower half object */
+static struct video_lowerhalf_s g_lower_half = {
+	&g_uvc_video_devops,		/* ops */
+	NULL						/* dev */
+};
+
+/* This is a bitmap that is used to allocate device names /dev/video<x>. */
+
+static uint32_t g_devinuse = 1;
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+* Name: uvc_get_colorspace
+*
+* Description:
+*   This is a helper function to return the supported V4L2 color format from
+*   UVC_VS_COLORFORMAT descriptor of the UVC.
+*
+* Input Parameters:
+*   clrindex - Color Index.
+*
+* Returned Values:
+*	 0 if no color format found, else return appropriate the V4L2 color format.
+*
+****************************************************************************/
+static uint32_t uvc_get_colorspace(const uint8_t clrindex)
+{
+	static const uint8_t primarycolor[] = {
+		0,
+		V4L2_COLORSPACE_SRGB,
+		V4L2_COLORSPACE_470_SYSTEM_M,
+		V4L2_COLORSPACE_470_SYSTEM_BG,
+		V4L2_COLORSPACE_SMPTE170M,
+		V4L2_COLORSPACE_SMPTE240M,
+	};
+
+	if (clrindex < ARRAY_SIZE(primarycolor)) {
+		return primarycolor[clrindex];
+	}
+
+	return 0;
```

Comment:
explicitly,
```
	static const uint8_t primarycolor[] = {
		V4L2_CORLORSPACE_NONE = 0,
		V4L2_COLORSPACE_SRGB,
		V4L2_COLORSPACE_470_SYSTEM_M,
		V4L2_COLORSPACE_470_SYSTEM_BG,
		V4L2_COLORSPACE_SMPTE170M,
		V4L2_COLORSPACE_SMPTE240M,
	};
	if (clrindex < ARRAY_SIZE(primarycolor)) {
		return primarycolor[clrindex];
	}

	return V4L2_CORLORSPACE_NONE;
```
but, format->colorspace using another place, so this structure seems to be moved to global area.

Reply:
uvc_get_colorspace is a static function, which is used only in this file. Here Primary color is an array holding the color space value which will return based on the index passed. I dont think we need to move this array out of this function scope.

Reply:
I don't think so, this returns one of primarycolor, so some of function will use this value.
so another function should refer to primarycolor...

Reply:
V4L2_COLORSPACE_xxx is in include/video/video.h which is accessible to all other files. Here we just fill the value to a static array to map it with index.

Reply:
ah.. then I understood.